### ISSUE TITLE
SEP V014: Modules and MapsTo

### ISSUE THREAD

**Issue by jakebeal** [SynBioDex/SBOL-visual#48]
The ModuleDefinition, Module, and MapsTo classes from SBOL 2 do not yet have an equivalent representation in SBOL Visual.

**Comment by jamesscottbrown** [comment]
The draft SEP is here: https://github.com/SynBioDex/SBOL-visual/blob/master/SEPs/SEP_V014.md

**Comment by jamesscottbrown** [comment]
After a discussion of a draft diagram, I have a couple of suggested additions:

* **Insert** "A port MUST NOT be connected to more than one glyph outside of the module, unless all of the glyphs it connects to represent the same entity"

* **After** "It is RECOMMENDED that a black-box module be drawn with ports, as if a line or arrow crosses the boundary of the module at a position other than a port it is impossible to see where it terminates.", **insert** "If a line or arrow crosses a module's boundary other than at a port, it SHOULD be labelled."

**Comment by jakebeal** [comment]
@jamesscottbrown Can you please edit these into the draft SEP?

**Comment by JS3xton** [comment]
Can you link to concrete definitions for Modules and the MapsTo relation? (sorry for being lazy and not digging it up myself)

How does the MapsTo relation relate to the Interaction Glyphs? (specifically, how is the MapsTo relation different than the Process Interaction Glyph?)

I also prefer not constraining the line style if possible (i.e. I don't like the idea of REQUIRING MapsTo lines to be dashed lines).

**Comment by cjmyers** [comment]
From the specification:

In particular, a MapsTo object provides two pieces of information: 

■ An identity relationship between two ComponentInstance objects,the first contained by the “lower level” definition of the ComponentInstance or Module that owns the MapsTo, and the second contained by the  “higher level” definition that contains the ComponentInstance or Module that owns the MapsTo. The remote  property of a MapsTo refers to the first “lower level” ComponentInstance, while the local property refers to  the second “higher level” ComponentInstance.

■ Instructions on how to interpret local and remote ComponentInstance objects that refer to different ComponentDefinition objects (that is, non-identical objects). These are specified using the refinement property of the MapsTo class. 

In other words, a MapsTo is used to indicate that the object in the containing module is the same as the one in the enclosed module.  Think of this like passing a parameter into a function in software.  A process, on the other hand, indicates that one object is transformed into another object.

I believe given that these are two very different concepts, it is important to distinguish them visually in some way.  One option could be via arrowhead, but there are not too many options available on this one, which is why we gravitated towards dashed lines.  Keep in mind though that this is not a requirement, but just a recommendation.  Other alternatives are provided as well to make them distinguished.

**Comment by jakebeal** [comment]
I've read through carefully, and am satisfied to see this move forward to a vote.

**Comment by JS3xton** [comment]
Some organizational suggestions for the current SEP:
- It might be helpful break out the specification of MapsTo (paragraphs 4-6 under the "2. Specification" heading) into a separate subsection (e.g. "Representation of MapsTo").
- It might be helpful to package the specification of "ports" into their own subsection or subsubsection.
- It might be helpful to package the special rules for "black-box" modules into their own subsection.
- It's unclear to me if: `"If a line or arrow crosses a module's boundary other than at a port, it SHOULD be labelled."` refers to *all* modules or just black-box modules.

Other comments:
- I don't know if I endorse RECOMMENDING ports on black-box modules.
- I want to amend my previous position on dashed lines for MapsTo (I was previously against RECOMMENDING them): I think they look pretty clean and work well in the left example.
- The following specification of MapsTo is confusing and seems to overload the Process Interaction Glyph:

    > A MapsTo relation MAY be indicated by a line with no arrowhead; alternatively, to visually distinguish between inputs and outputs of a module, a MapsTo relation for the input of a module MAY be drawn with a filled arrowhead directed into the module, and a MapsTo relation for the output of a module MAY be drawn with a filled arrowhead directed out of the module.
    >
    > As an alternative to depicting MapsTo relations, it is permitted to draw arrows entering a module. In such a case, the representation of the module does not modify the semantics of the interactions shown: it simply indicates that where the corresponding interaction is defined.

    Here are all the ways I can think of to draw SBOL Visual arrows relative to a module boundary:

    ![arrow_module](https://user-images.githubusercontent.com/2179825/46783236-1a94a780-ccef-11e8-9998-cded45a139b9.png)

    - `A MapsTo relation MAY be indicated by a line with no arrowhead` (**1,7,13,19**)
    - `a MapsTo relation for the input of a module MAY be drawn with a filled arrowhead directed into the module` (**2,8**. Overloads Process Interaction Glyph)
    - `a MapsTo relation for the output of a module MAY be drawn with a filled arrowhead directed out of the module` (**14,20**. Overloads Process Interaction Glyph)
    - (*)`As an alternative to depicting MapsTo relations, it is permitted to draw arrows entering a module. In such a case, the representation of the module does not modify the semantics of the interactions shown: it simply indicates that where the corresponding interaction is defined.` - I don't understand what this means.
    - Everything else, by default, MUST be an Interaction Glyph? (and thus not a MapsTo). It's unclear how those interactions are modified by the module, though (one would hope intuitively, i.e. the interaction sources and sinks exist inside or outside of the module as appropriate). Perhaps this is what (*) was trying to get at.

    I'm also actually fine with overloading the Process Glyph, I just think it needs to be clearly specified.

- I think the mention of `ModuleDefinition` (which is an SBOL concept) at the end of the "2. Specification" section should be replaced with "module" (which we are defining here as an SBOL Visual concept).
- On the left example, are the arrows from cas9_generic and gRNA_generic to cas9_gRNA_complex MapsTo arrows? (I don't think they should be, but they're dotted lines, which is a little confusing):

    ![image](https://user-images.githubusercontent.com/2179825/46784158-a1974f00-ccf2-11e8-91a6-fa878a9aeaaa.png)

- On the right example, the arrow colors are a little confusing for me. I assume it's supposed to indicate Interactions that take place inside the module? At first I thought it was simply the "distinguishing feature" that differentiated Process Interaction Glyphs from MapsTo glyphs, but upon studying the diagram, the Process Interaction Glyphs are both black and purple whereas the MapsTo glyphs are purple. (Unrelated: the stimulation glyph is also used across a module boundary, which, as noted above, is arguably not well defined).

**Comment by jakebeal** [comment]
There are some good catches there, John.

> It's unclear to me if: "If a line or arrow crosses a module's boundary other than at a port, it SHOULD be labelled." refers to all modules or just black-box modules.

Re-reading this, I am unclear as well, particularly because this isn't shown in any of our examples (and they seem to be clear).  I think this sentence should either be embodied in examples or else removed.

> I don't know if I endorse RECOMMENDING ports on black-box modules.

I can see it either way: I think it would be good to have "RECOMMENDED" vs. "OPTIONAL" be a choice in the adoption vote.

On process glyph overloading:
> a MapsTo relation for the (input/output) of a module MAY be drawn with a filled arrowhead directed into the module

Looking at our examples, it seems that we only draw arrows if we're showing the MapsTo+interaction combined, rather than showing the MapsTo separately.  Thus, I think it makes sense to either remove this overloading option or else add examples that really need it.

> On the left example, are the arrows from cas9_generic and gRNA_generic to cas9_gRNA_complex MapsTo arrows? 

These are showing an interaction, and thus should be solid --- and also should probably be replaced with an association multi-tail arrow per SEP V013, since that looks likely to be adopted.

**Comment by JS3xton** [comment]
> On process glyph overloading:
>
> > a MapsTo relation for the (input/output) of a module MAY be drawn with a filled arrowhead directed into the module
>
> Looking at our examples, it seems that we only draw arrows if we're showing the MapsTo+interaction combined, rather than showing the MapsTo separately. Thus, I think it makes sense to either remove this overloading option or else add examples that really need it.

When I mentioned "overloading", I suppose I meant that an arrow could be *either* a MapsTo *or* an Interaction glyph, and which one it is would be based on context. You seem to be proposing that an arrow could be *both* a MapsTo *and* an Interaction glyph. This makes a lot of sense to me. Again, I would encourage that we specify it concretely, though. Perhaps: "any Interaction Glyph that crosses a module boundary implies an unillustrated MapsTo glyph". Some examples:
<br>

![arrow_module2](https://user-images.githubusercontent.com/2179825/46821106-73932880-cd4d-11e8-9754-21ade5482bcb.png)

This use case is embodied by the black solid-head arrows in the Right Example (are those Process glyphs? MapsTo glyphs? Both?).

It follows that to specify *only* the MapsTo relationship, you'd have to use *NOT* an Interaction Glyph (i.e. a no-head arrow? With a RECOMMENDED line style of dashed?).

Also, how do modules relate to membranes (i.e. "physical" modules)? Can membranes simply be classified as modules? There are additional membrane details (e.g. proximity to and structure of) that may not hold generally for modules.

**Comment by jakebeal** [comment]
@JS3xton I agree with your interpretation and examples regarding MapsTo.

With respect to membranes: currently, the specification is silent on this question.  I think in many cases they would make sense to specify as modules, but would suggest deferring a careful consideration of them for later.

**Comment by cjmyers** [comment]
In SBML, membranes are handled separately from modules.  Namely, membranes are represented as compartments while modules are subModels.  I agree with Jake that we have not settled this one, and I can see a point for not conflating physical boundaries with simply hierarchical representation of a design.

**Comment by jakebeal** [comment]
Based on these conversations, I have prepared an updated version of SEP V013, along with draft changes to the spec. The pull request containing the full diffs is here: #56

The updated SEP V014 in particular may be found here: https://github.com/SynBioDex/SBOL-visual/blob/c0445cf/SEPs/SEP_V014.md

**Comment by jakebeal** [comment]
Accepted and integrated, and thus closed per SBOL procedure in updated SEP 001.
